src/routing_table/mod.rs

```diff
@@ -605,6 +605,40 @@ impl<T: Binary + Clone + Copy + Debug + Default + Hash + Xorable> RoutingTable<T
             .collect()
     }
 
+    /// Returns the `route`-th node in the given group, sorted by distance to `target`
+    pub fn get_routeth_node(&self,
+                            group: &HashSet<T>,
+                            target: T,
+                            exclude: Option<T>,
+                            route: usize)
+                            -> Result<T, Error> {
+        let mut names = if let Some(exclude) = exclude {
+            group.iter().filter(|&x| *x != exclude).collect_vec()
+        } else {
+            group.iter().collect_vec()
+        };
+        names.sort_by(|&lhs, &rhs| target.cmp_distance(lhs, rhs));
+        match names.get(route) {
+            Some(&name) => Ok(*name),
+            None => {
+                // TODO: This is a workaround for the cases where we have not connected to all
+                //       needed contacts yet and may have empty or incomplete groups.
+                // Err(Error::CannotRoute),
+                let cmp = |name0: &T, name1: &T| target.cmp_distance(name0, name1);
+                match self.groups
+                    .iter()
+                    .map(|(prefix, group)| (prefix.lower_bound(), group))
+                    .sorted_by(|&(name0, _), &(name1, _)| cmp(&name0, &name1))
+                    .into_iter()
+                    .flat_map(|(_, peers)| peers.iter().cloned().sorted_by(&cmp).into_iter())
+                    .next() {
```

Comment:
TBH, I just moved this code around and haven't analysed it too much. 

Wrapping around would make some sense, but we would then retry routes that already failed before giving up, so I'm not sure if that's a good idea. Maybe we should just fail when we run out of names.